[Navigation] Implement NavigationAPIMethodTracker

TingPing opened the pull request:
#### 5592c3df7e5002c0018419e8b8694341206ba7ad
<pre>
[Navigation] Implement NavigationAPIMethodTracker
<a href="https://bugs.webkit.org/show_bug.cgi?id=269930">https://bugs.webkit.org/show_bug.cgi?id=269930</a>

Reviewed by Alex Christensen.

This creates a NavigationAPIMethodTracker struct per:
<a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker">https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker</a>

As well as the maybeSetUpcomingNonTraversalTracker() method per:
<a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#maybe-set-the-upcoming-non-traverse-api-method-tracker">https://html.spec.whatwg.org/multipage/nav-history-apis.html#maybe-set-the-upcoming-non-traverse-api-method-tracker</a>

* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::serializeState):
(WebCore::Navigation::maybeSetUpcomingNonTraversalTracker):
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
* Source/WebCore/page/Navigation.h:
(WebCore::NavigationAPIMethodTracker::create):
(WebCore::NavigationAPIMethodTracker::NavigationAPIMethodTracker):

Canonical link: <a href="https://commits.webkit.org/275446@main">https://commits.webkit.org/275446@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be820df77fdc0f977e9f9de5ea79e63e7bf7cbf7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44433 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37946 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44158 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24021 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18196 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42425 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17796 "Found 2 new test failures: fast/forms/datalist/datalist-option-labels.html, fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36041 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15260 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15484 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45834 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38040 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37386 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41147 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16666 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13686 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18285 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18344 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5610 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17929 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -127,16 +127,52 @@ static Navigation::Result createErrorResult(Ref<DeferredPromise> committed, Ref<
+{
+    auto globalObject = committed->globalObject();
+    Ref committedPromise = DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(committed->promise()));
+    Ref finishedPromise = DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(finished->promise()));
```
These should probably be properly moved into NavigationAPIMethodTracker::create or NavigationAPIMethodTracker should create them itself (based on the two DeferredPromise params).

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -127,16 +127,48 @@ static Navigation::Result createErrorResult(Ref<DeferredPromise> committed, Ref<
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#maybe-set-the-upcoming-non-traverse-api-method-tracker
+Ref<NavigationAPIMethodTracker> Navigation::maybeSetUpcomingNonTraversalTracker(Ref<DeferredPromise>& committed, Ref<DeferredPromise>& finished, JSC::JSValue&& info, std::optional<Ref<SerializedScriptValue>>&& serializedState)
+{
+    Ref apiMethodTracker = NavigationAPIMethodTracker::create(committed, finished, WTFMove(info), serializedState);
```
Do we want NavigationAPIMethodTracker to take ownership of the promises? Then we can use move semantics to make that a bit clearer.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -161,23 +193,26 @@ Navigation::Result Navigation::navigate(ScriptExecutionContext& scriptExecutionC
+        m_upcomingNonTraverseMethodTracker = std::nullopt;
+    }
+
     // FIXME: keep track of promises to resolve later.
```
Is this the job of NavigationAPIMethodTracker? If so I think the FIXME can be removed.

TingPing replied:
Yes but not everything is done to resolve the promises at the right moment yet.

achristensen07 reviewed (commented):
Tests?

achristensen07 commented on Source/WebCore/page/Navigation.h:
```diff
@@ -104,10 +130,16 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
     RefPtr<NavigationTransition> m_transition;
     Vector<Ref<NavigationHistoryEntry>> m_entries;
+
+    std::optional<Ref<NavigationAPIMethodTracker>> m_upcomingNonTraverseMethodTracker;
```
optional<Ref> might be able to be RefPtr

darinadler commented on Source/WebCore/page/Navigation.h:
```diff
@@ -31,10 +31,36 @@
+class SerializedScriptValue;
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker
+struct NavigationAPIMethodTracker : public RefCounted<NavigationAPIMethodTracker> {
```
Why is this a reference-counted object instead of just a `struct`?

TingPing replied:
I felt like this is just the nicer option.

I tried `unique_ptr` but that doesn't work for its usage as by spec it will have multiple places referencing it.

It seems unnecessary to ever copy it as a plain instance.

So I think either its a raw pointer or a `RefPtr` is cleaner.

darinadler replied:
That sounds wrong to me. Seems to me that we should be passing and moving an object around, not allocating it on the heap.

TingPing replied:
Ok done.

darinadler commented on Source/WebCore/page/Navigation.h:
```diff
@@ -31,10 +31,36 @@
+        return adoptRef(new NavigationAPIMethodTracker(WTFMove(committed), WTFMove(finished), WTFMove(info), WTFMove(serializedState)));
+    };
+
+    std::optional<String> key;
```
Since a `String` has a null value, we don’t need to use `std::optional`.

darinadler commented on Source/WebCore/page/Navigation.h:
```diff
@@ -31,10 +31,36 @@
+    std::optional<String> key;
+    JSC::JSValue info;
+    std::optional<Ref<SerializedScriptValue>> serializedState;
+    std::optional<Ref<NavigationHistoryEntry>> committedToEntry;
```
We can use `RefPtr` instead of `std::optional<Ref>`.

darinadler commented on Source/WebCore/page/Navigation.h:
```diff
@@ -31,10 +31,38 @@
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker
+struct NavigationAPIMethodTracker {
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
```
Not sure we need this, since we don’t allocate this on the heap.

darinadler commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -161,23 +193,26 @@ Navigation::Result Navigation::navigate(ScriptExecutionContext& scriptExecutionC
     window()->frame()->loader().load(FrameLoadRequest(*window()->frame(), newURL));
 
+    if (m_upcomingNonTraverseMethodTracker && m_upcomingNonTraverseMethodTracker.value() == apiMethodTracker) {
+        // TODO: Once the frameloader properly promotes the upcoming tracker with the navigate event `m_upcomingNonTraverseMethodTracker` should be unset or this will throw.
```
WebKit project uses `FIXME`.

darinadler commented on Source/WebCore/page/Navigation.h:
```diff
@@ -104,10 +132,16 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
     bool hasEntriesAndEventsDisabled() const;
     Result performTraversal(NavigationHistoryEntry&, Ref<DeferredPromise> committed, Ref<DeferredPromise> finished);
     std::optional<Ref<NavigationHistoryEntry>> findEntryByKey(const String& key);
+    ExceptionOr<RefPtr<SerializedScriptValue>> serializeState(JSC::JSValue& state);
```
Argument should just be `JSValue`, not `JSValue&`.

darinadler commented on Source/WebCore/page/Navigation.h:
```diff
@@ -104,10 +132,16 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
     Result performTraversal(NavigationHistoryEntry&, Ref<DeferredPromise> committed, Ref<DeferredPromise> finished);
     std::optional<Ref<NavigationHistoryEntry>> findEntryByKey(const String& key);
+    ExceptionOr<RefPtr<SerializedScriptValue>> serializeState(JSC::JSValue& state);
+    NavigationAPIMethodTracker maybeSetUpcomingNonTraversalTracker(Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished, JSC::JSValue&& info, RefPtr<SerializedScriptValue>&&);
```
Third argument should just be `JSValue`, not `JSValue&&`.

TingPing committed:
[Navigation] Implement NavigationAPIMethodTracker
https://bugs.webkit.org/show_bug.cgi?id=269930

Reviewed by Alex Christensen.

This creates a NavigationAPIMethodTracker struct per:
https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-method-tracker

As well as the maybeSetUpcomingNonTraversalTracker() method per:
https://html.spec.whatwg.org/multipage/nav-history-apis.html#maybe-set-the-upcoming-non-traverse-api-method-tracker

* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::serializeState):
(WebCore::Navigation::maybeSetUpcomingNonTraversalTracker):
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
* Source/WebCore/page/Navigation.h:
(WebCore::NavigationAPIMethodTracker::create):
(WebCore::NavigationAPIMethodTracker::NavigationAPIMethodTracker):

Canonical link: https://commits.webkit.org/275446@main

darinadler commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -161,23 +194,26 @@ Navigation::Result Navigation::navigate(ScriptExecutionContext& scriptExecutionC
     // FIXME: This is not a proper Navigation API initiated traversal, just a simple load for now.
     window()->frame()->loader().load(FrameLoadRequest(*window()->frame(), newURL));
 
+    if (m_upcomingNonTraverseMethodTracker && m_upcomingNonTraverseMethodTracker.value() == apiMethodTracker) {
```
Can this just be like this?
```
if (m_upcomingNonTraverseMethodTracker == apiMethodTracker)
```
I think `std::optional` works like this.

darinadler reviewed (commented):
I had review comments earlier but I didn’t get them in before you landed.
